zerver/tests/test_invite.py

```diff
@@ -3118,6 +3449,95 @@ def test_create_multiuse_link_with_specified_user_groups_api_call(self) -> None:
+        realm = get_realm("zulip")
+        hamlet = self.example_user("hamlet")
+        iago = self.example_user("iago")
+        cordelia = self.example_user("cordelia")
+
+        stream_names = ["Denmark", "Scotland"]
+        streams = [get_stream(stream_name, realm) for stream_name in stream_names]
+        stream_ids = [stream.id for stream in streams]
+
+        multiuse_invite = MultiuseInvite.objects.create(
+            referred_by=self.example_user("iago"),
+            realm=realm,
+        )
+        multiuse_invite.streams.set(stream_ids)
+
+        def check_path_multiuse_invite_with_subscribe_permission(
+            user: UserProfile,
+            stream_ids: list[int],
+            allow_fail: bool = False,
+            *,
+            error_message: str | None = None,
+        ) -> None:
+            result = self.api_patch(
+                user,
+                f"/api/v1/invites/multiuse/{multiuse_invite.id}",
+                dict(
+                    stream_ids=orjson.dumps(stream_ids).decode(),
+                    include_realm_default_subscriptions=orjson.dumps(False).decode(),
+                ),
+            )
+            if allow_fail and error_message is not None:
+                self.assert_json_error(result, error_message)
+            else:
+                self.assert_json_success(result)
+
+            multiuse_invite.refresh_from_db()
+            self.assertEqual(
+                set(multiuse_invite.streams.values_list("id", flat=True)),
+                set(stream_ids),
+            )
+
+        admins_group = NamedUserGroup.objects.get(
+            name=SystemGroups.ADMINISTRATORS, realm=realm, is_system_group=True
+        )
+        nobody_group = NamedUserGroup.objects.get(
+            name=SystemGroups.NOBODY, realm=realm, is_system_group=True
+        )
+
+        # Realm-level: only admins can add subscribers
+        do_change_realm_permission_group_setting(
+            realm, "can_add_subscribers_group", admins_group, acting_user=None
+        )
+
+        check_path_multiuse_invite_with_subscribe_permission(
+            hamlet, stream_ids, True, error_message="Insufficient permission"
+        )
+
+        check_path_multiuse_invite_with_subscribe_permission(iago, stream_ids)
+
+        do_change_realm_permission_group_setting(
+            realm, "can_add_subscribers_group", nobody_group, acting_user=None
+        )
+
+        check_path_multiuse_invite_with_subscribe_permission(iago, stream_ids)
+
+        # This anonymous group contains only hamlet.
+        anonymous_group = UserGroupMembersData(direct_members=[hamlet.id], direct_subgroups=[])
+
+        do_change_stream_group_based_setting(
+            streams[0],
+            "can_add_subscribers_group",
+            anonymous_group,
+            acting_user=cordelia,
+        )
+
+        # Hamlet can edit for Denmark only
+        check_path_multiuse_invite_with_subscribe_permission(hamlet, [streams[0].id])
```

Comment:
If you see the error traceback, this fails because user does not have permission to create multiuse invite. And the above tests for failure cases are also incorrect because those are either failing for not having permission to create or non-admin trying to edit invites created by someone else. I guess we do not have "Insufficient permission" error message when user does not have permission to subscribe. So, make sure we are testing the correct failure case.